### PR TITLE
Performance improvements

### PR DESCRIPTION
**Pull request by wezell**

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java [hunk 1]
```diff
@@ -963,7 +966,45 @@ private IHTMLPage loadPageByIdentifier ( String ident, boolean live, User user,
...
+      }else if(theField instanceof CategoryField){
+        Category category;
+        try {
+          category = catAPI.find(theField.values(), APILocator.getUserAPI().getSystemUser(), false);
+          // Get all the Contentlets Categories
+          List<Category> selectedCategories = catAPI.getParents(contentlet, APILocator.getUserAPI().getSystemUser(), false);
+          Set<Category> categoryList = new HashSet<Category>();
+          List<Category> categoryTree = catAPI.getAllChildren(category, APILocator.getUserAPI().getSystemUser(), false);
+          if (selectedCategories.size() > 0 && categoryTree != null) {
+              for (int k = 0; k < categoryTree.size(); k++) {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java]: If we use a Set we can reduce the cyclic complexity, such as 

Set<Category> categoryTree = new TreeSet ( (cat1, cat3) -> categ.getInode().compareTo(cat.getInode()))
for (Category categ : selectedCategories) {
                    if (categoryTree.contains (categ) {
                         categoryList.add(categoryTree.get(categ));
                      }
                  }

File: dotCMS/src/main/java/com/dotmarketing/business/APILocator.java [hunk 2]
```diff
@@ -337,7 +337,10 @@ public static RelationshipAPI getRelationshipAPI(){
 	public static com.dotmarketing.portlets.structure.business.FieldAPI getFieldAPI(){
 		return (com.dotmarketing.portlets.structure.business.FieldAPI)getInstance(APIIndex.FIELD_API);
 	}
-
+	
+    public static com.dotcms.contenttype.business.FieldAPI  getFieldAPI2(){
```
Comment (jgambarios) [inline dotCMS/src/main/java/com/dotmarketing/business/APILocator.java]: I think we need a better name for this method

Comment (jgambarios) [inline dotCMS/src/main/java/com/dotmarketing/business/APILocator.java]: I think we need a better name for this method

File: dotCMS/src/main/java/com/dotmarketing/portlets/contentlet/model/Contentlet.java [hunk 3]
```diff
@@ -182,7 +183,17 @@ public Structure getStructure() {
     	structure = CacheLocator.getContentTypeCache().getStructureByInode(getStructureInode());
         return structure;
     }
+    
+    /**
+     */
+    public ContentType getContentType() {
```
Comment (jgambarios) [inline dotCMS/src/main/java/com/dotmarketing/portlets/contentlet/model/Contentlet.java]: I think we should avoid to add business logic into POJOs

**Review by jgambarios — APPROVED** [review]

File: dotCMS/src/main/java/com/dotmarketing/portlets/contentlet/model/Contentlet.java (hunk 3 above)
Comment (wezell) [inline dotCMS/src/main/java/com/dotmarketing/portlets/contentlet/model/Contentlet.java]: This was done to mirror existing functionality, but you are right.  Removed and deprecated old functionality.

File: dotCMS/src/main/java/com/dotmarketing/business/APILocator.java (hunk 2 above)
Comment (wezell) [inline dotCMS/src/main/java/com/dotmarketing/business/APILocator.java]: this is open to suggestions....

getNewFieldAPI?  getNFieldAPI?

Comment (jcastro-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/business/APILocator.java]: @wezell what about "getContentTypeFieldAPI" ?

Comment (wezell) [inline dotCMS/src/main/java/com/dotmarketing/business/APILocator.java]: Well, Jose,  your suggestion was so good that it was already implemented like that.  Turns out we don't even need the getFieldAPI2  method.

**Pull request merged**